### ISSUE TITLE
Expose a `gc` method in dart:developer

### ISSUE THREAD

**Issue by jacob314** [dart-lang/sdk#45458]
Add a `dart:developer` method that forces a GC.
The method is only allowed (fine for it to even throw an exception) if the Dart VM is launched with a specific flag which is consistent with how Chrome only exposes `window.gc` if
`--js-flags=--expose-gc`
is passed.
Providing this option will make it easier to write tests that verify that objects do not leak.

**Comment by jacob314** [comment]
CC @lrhn @natebosch @srujzs @sigmundch @nshahan @bkonyi

**Comment by natebosch** [comment]
I'd give it a longer method name, but I think it would be fine to have this API as long as the implementations can support.

**Comment by jacob314** [comment]
maybe`debugForceGc()` to clarify that the method is only available in debug builds and users should not count on it being available in release builds?

**Comment by natebosch** [comment]
Maybe `forceGarbageCollection()`?

**Comment by sigmundch** [comment]
> ...  to clarify that the method is only available in debug builds and users should not count on it being available in release builds?

should anything in `dart:developer` be considered available for production/release use? Maybe we can address that concern more coarsely with the entire library?

**Comment by jacob314** [comment]
Good point. Almost everything in `dart:developer` is at minimum useless in production/release. There are a few utility classes that work equally well in production but all the core functionality like dispatching timeline events won't do anything.

**Comment by a-siva** [comment]
In most cases such a method would be a no op in the VM because of the way garbage collection is implemented in the VM. At best it would be equivalent to the idle notification mechanism that flutter engine uses.

**Comment by nshahan** [comment]
@jacob314 Would this API addition to dart:developer carry its weight if it is a no op in the VM? Still worth the addition only for development on the web platform?

**Comment by jacob314** [comment]
It is only useful if it works on both. @a-siva, the button to force GC in the Observatory and DevTools seems like it does something useful when I used it with my prototype leak detection tool. I'd see disposed objects that were GCed as soon as I clicked the GC button but maybe I was just confusing myself.

**Comment by dnfield** [comment]
On the VM, it would probably be better if we just had some way to force a call to `Dart_NotifyLowMemory`, which AFAIK will more or less force the most aggressive GC possible. @a-siva is that not the case?

**Comment by dnfield** [comment]
(Signalling an actual platform low memory to the app would additionally cause Flutter to notify the application of low memory pressure, which would clear the image cache and potentially other caches in the app if it's listening for that signal.)

**Comment by dnfield** [comment]
Well, I was kind of expecting someone else to say these things, but here are a bunch of reasons we _shouldn't_ do this:

- Developers will treat this like a magic memory management tool. Even if we enable it only for development builds, they'll get confused about how, why, and whether it works, and also be confused/upset that it doesn't work in release. It will get sprinkled around code bases.
- If we enable it in multiple modes, small differences between debug/release builds will confuse developers.
- Developers will be confused about differences in behavior across web and VM.
- Developers will be confused about why this doesn't do anything sometimes even though they can prove there's unreachable garbage (e.g. because a heuristic decides to defer collecting).

Is there some other more useful API we can expose? Or, is there an existing API we can use? For example, the observatory currently shows "uncollected garbage" when doing heap snapshots. I could imagine it would be useful if you could pass a weak reference to the VM and ask if it's collectible or not.

**Comment by jacob314** [comment]
Good points. 
The only API I really care about is a query for whether a weak reference (or more efficiently, each reference in a list of weak references) is collectable or not. That API seems less likely to be abused by users as it couldn't be mistaken for a memory management tool. 
One downside is Dart doesn't surface weak references directly so I'm not sure how to express that API in Dart without adding a WeakRef class to Dart.

**Comment by dnfield** [comment]
Even something like being able to link up an object key of some sort (maybe identityHashCode?) to an item in the "uncollected garbage" portion of the heap map would probably do that, right?

I could imagine something like a VM service call that you give a list of identity hash codes to, and it does the work of creating a heapsnapshot and combing through for those hashcodes in the uncollected garbage portion.

Obviously this wouldn't work for web...  But as a VM service method it would be clear this is VM only. /cc @rmacnak-google @bkonyi WDYT?

**Comment by nshahan** [comment]
> Obviously this wouldn't work for web... But as a VM service method it would be clear this is VM only. 

I just want to point out that just being part of the VM service protocol doesn't exactly make it clear it is VM only. The infra between the browser and the dev tools translates to the VM service protocol in an effort to make the same debug tools work for web debugging.

cc @grouma @annagrin

**Comment by grouma** [comment]
> I just want to point out that just being part of the VM service protocol doesn't exactly make it clear it is VM only. The infra between the browser and the dev tools translates to the VM service protocol in an effort to make the same debug tools work for web debugging.

There are existing portions of the VM service protocol that don't apply to the web platform so I don't think it's much of a concern. It would be good to update the protocol documentation to make it explicit however.
